iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs

```diff
@@ -522,19 +519,27 @@ private async Task DisposeMessageAsync(string lockToken, AmqpIoTDisposeActions o
 
         private void TwinMessageListener(Twin twin, string correlationId, TwinCollection twinCollection)
         {
-            if (correlationId != null)
+            if (correlationId == null)
             {
-                // It is a GET, just complete the task.
-                TaskCompletionSource<Twin> task;
-                if (_twinResponseCompletions.TryRemove(correlationId, out task))
-                {
-                    task.SetResult(twin);
-                }
+                // This is desired property updates, so call the callback with TwinCollection.
+                _onDesiredStatePatchListener(twinCollection);
             }
             else
             {
-                // It is a PATCH, just call the callback with the TwinCollection
-                _onDesiredStatePatchListener(twinCollection);
+                if (correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase) ||
+                    correlationId.StartsWith(AmqpTwinMessageType.Patch.ToString(), StringComparison.OrdinalIgnoreCase))
+                {
+                    // For Get and Patch, complete the task.
+                    TaskCompletionSource<Twin> task;
+                    if (_twinResponseCompletions.TryRemove(correlationId, out task))
+                    {
+                        task.SetResult(twin);
+                    }
+                    else
+                    {
+                        Logging.Info("Could not remove a correlation Id from a dictionary", nameof(TwinMessageListener));
```

Comment:
how about rewording this to explain why the TryRemove could fail, something along the lines of - 
"attempted to complete the task awaiter for a twin operation whose request did not originate from the sdk - correlation ID was not set by the sdk"?